oap-server/analyzer/log-analyzer/src/main/java/org/apache/skywalking/oap/log/analyzer/provider/log/listener/LogFilterListener.java

```
@@ -19,61 +19,77 @@
 
     @Override
     public void build() {
-        dsls.forEach(dsl -> {
-            try {
-                dsl.evaluate();
-            } catch (final Exception e) {
-                log.warn("Failed to evaluate dsl: {}", dsl, e);
-            }
-        });
+        try {
+            dsl.evaluate();
+        } catch (final Exception e) {
+            log.warn("Failed to evaluate dsl: {}", dsl, e);
+        }
     }
 
     @Override
     public LogAnalysisListener parse(final LogData.Builder logData,
                                      final Message extraLog) {
-        dsls.forEach(dsl -> dsl.bind(new Binding().log(logData.build())
-                                                  .extraLog(extraLog)));
+        dsl.bind(new Binding().log(logData.build()).extraLog(extraLog));
         return this;
     }
 
     public static class Factory implements LogAnalysisListenerFactory {
-        private final List<DSL> dsls;
+        private final Map<Layer, DSL> dsls;
 
         public Factory(final ModuleManager moduleManager, final LogAnalyzerModuleConfig config) throws Exception {
-            dsls = new ArrayList<>();
+            dsls = new HashMap<>();
 
             final List<LALConfig> configList = LALConfigs.load(config.getLalPath(), config.lalFiles())
                                                          .stream()
                                                          .flatMap(it -> it.getRules().stream())
                                                          .collect(Collectors.toList());
             for (final LALConfig c : configList) {
-                dsls.add(DSL.of(moduleManager, config, c.getDsl()));
+                try {
+                    Layer layer = Layer.nameOf(c.getLayer());
+                    if (dsls.put(layer, DSL.of(moduleManager, config, c.getDsl())) != null) {
+                        log.warn("layer {} has already set a rule {}, the old one will be ignored.", layer, c.getName());
+                    }
```

Layer conflicts should cause booting failure. Please throw the exception with clear information about which rules(files) are using the same layer.

---

DSL has not stored the name of rule. So it is difficult to log which rules(files) are using the same layer.

---

> DSL has not stored the name of rule. So it is difficult to log which rules(files) are using the same layer.

Simply log the layer should be enough IMO